Correctness/Correctness.lean

```diff
@@ -175,4 +215,431 @@ theorem partial_correctness_from_verification_conditions [Sys σ] [Spec' σ]
         ⟨n, Nat.le_refl _, hexit, hpost⟩
     find 0 (v1 s0 hp) (Nat.zero_le ..)
 
+----------------------------------------------------------------------
+
+-- Prove partial correctness using inductive assertions using the
+-- function `cassert` that checks if `assert` holds if at a cutpoint
+-- state, or else recurs until it hits one. This function is also
+-- partial: if no cutpoint is reachable, the recursion doesn't
+-- terminate.
+
+/--
+`cassert s0 si i`: if this function terminates, it returns a pair
+whose first element is a `Nat` that is `(i + the number of steps simulated
+from si to reach the next cutpoint)`, and the second element is a
+`Prop` that checks whether `assert` holds at that cutpoint state.
+-/
+noncomputable def cassert [Sys σ] [Spec' σ] (s0 si : σ) (i : Nat) : Nat × Prop :=
+  iterate
+    (fun (si, i) =>
+        if cut si then
+          .inl (i, assert s0 si)
+        else
+          .inr (next si, i + 1))
+    (si, i)
+
+theorem cassert_eq [Sys σ] [Spec' σ] (s0 si : σ) (i : Nat)
+        : cassert s0 si i = if cut si then (i, assert s0 si)
+                       else cassert s0 (next si) (i + 1) := by
+  unfold cassert
+  conv => lhs; rw [iterate_eq]
+  by_cases cut si <;> simp only [Bool.false_eq_true, ↓reduceIte, *]
+  done
+
+theorem cassert_cut [Sys σ] [Spec' σ] {s0 si : σ} (h : cut si) (i : Nat) :
+  (cassert s0 si i).fst = i ∧
+  (cassert s0 si i).snd = assert s0 si := by
```

Comment:
```suggestion
    (cassert s0 si i).fst = i ∧
    (cassert s0 si i).snd = assert s0 si := by
```
Not sure whether we want to keep following the style we currently have (which has the statement and proof at the same indentation), or slowly migrate toward standard Lean style (which has the statement indented one more level than the proof, as in my suggestion)